### PR TITLE
MOTOR DRV: Add motor configs

### PR DESCRIPTION
**Pull request by StephanCoetzee**
The new motor drivers spin the motors as expected.

@Jeff-Ciesielski

### CODE REVIEW

File: cpu_map_keyme2560.h [hunk 1]
```diff
@@ -70,24 +61,36 @@
...
+#define STEP_DDR      DDRH
+#define STEP_PORT     PORTH
+#define STEP_PIN      PINH
+#define X_STEP_BIT    PH0
+#define Y_STEP_BIT    PH1
+#define Z_STEP_BIT    PH2
+#define C_STEP_BIT    PH3
+/* All step bits */
+#define STEP_MASK     ((1 << X_STEP_BIT) | (1 << Y_STEP_BIT) | (1 << Z_STEP_BIT) | (1 << C_STEP_BIT))
+
```
Comment (Jeff-Ciesielski) [inline cpu_map_keyme2560.h:74]: Why move this?

File: motor_driver.c [hunk 2]
```diff
@@ -127,19 +145,65 @@ void motor_drv_disable_motor(enum stepper_e stepper)
...
 
-void motor_drv_init()
+void motor_drv_inverse_dir(enum stepper_e stepper, bool reversed)
 {
-  /* Configure reset pin as output */
-  MOTOR_RESET_DDR |= ((1 << MOTOR_RESET_PIN) | (1 << MOTOR_RESET_LINE_DRIVER_PIN));
+  if (reversed) {
+    _motor_drv_set_val(stepper, CTRL, RDIR_IDX, RDIR_MASK, 1);
+  } else {
+    _motor_drv_set_val(stepper, CTRL, RDIR_IDX, RDIR_MASK, 1);
```
Comment (Jeff-Ciesielski) [inline motor_driver.c]: I'm pretty sure this will do the same thing no matter what reversed value you pass?

File: stepper.c [hunk 3]
```diff
@@ -236,9 +232,19 @@ static uint16_t st_shutdown_delay;  //ms (max = 32767)
...
-  else { STEPPERS_DISABLE_PORT &= ~(STEPPERS_DISABLE_MASK&mask); }
+  if (bit_istrue(settings.flags, BITFLAG_INVERT_ST_ENABLE)) {
+    disable = !disable;  // Apply pin invert.
+  }
+
+  if (disable) {
+    STEPPERS_DISABLE_PORT |= (STEPPERS_DISABLE_MASK & mask);
+  } else { 
+    STEPPERS_DISABLE_PORT &= ~(STEPPERS_DISABLE_MASK & mask);
+  }
```
Comment (Jeff-Ciesielski) [inline stepper.c]: Did anything in the disable logic change?  Mild preference to revert it if the logic is the same (or move it into its own commit at least).

Comment (Jeff-Ciesielski) [reply stepper.c]: FWIW: I agree that this is a good improvement.

File: motor_driver.c [hunk 4]
```diff
@@ -127,19 +145,65 @@ void motor_drv_disable_motor(enum stepper_e stepper)
...
+
+  motor_drv_set_torque(CAROUSEL, TORQUE_VAL_5A);
+  motor_drv_set_isgain(CAROUSEL, FIVE);
+
+  motor_drv_set_torque(GRIPPER, TORQUE_VAL_3A);
+  motor_drv_set_isgain(GRIPPER, FIVE);
+
+  for (int idx = 0; idx < 4; idx++) {
+    /* Set microstepping */
+    uint8_t mask = settings.microsteps & (0x3 << (2 * idx));
```
Comment (Jeff-Ciesielski) [inline motor_driver.c]: const

Comment (Jeff-Ciesielski) [reply motor_driver.c]: Also, this feels like it could be its own helper function maybe?

File: motor_driver.c [hunk 5]
```diff
@@ -20,64 +21,81 @@
...
+  /* Read-modify-write the specified address */
 
-  data = data_array[0] | (data_array[1] << 8);
+  /* Read the register */
+  uint16_t data;
+  data = _motor_drv_read_reg(stepper, address);
 
   /* Clear the bits that need to be set */
-  data &= (mask << idx);
+  data &= ~(mask << idx);
```
Comment (Jeff-Ciesielski) [inline motor_driver.c:99]: :+1:

**Review by Jeff-Ciesielski — COMMENTED** [review]
Glad these puppies are spinning :D.  Some small fixes.

File: motor_driver.c (hunk 2 above)
Comment (StephanCoetzee) [reply motor_driver.c]: That's a typo. Thanks

File: stepper.c (hunk 3 above)
Comment (StephanCoetzee) [reply stepper.c]: The logic is the same. I changed it because I was reviewing this piece of code while I was changing something in the eeprom settings and I didn't want to deal with reading those horrible one-liners.

I agree with reverting it for this PR.

File: cpu_map_keyme2560.h (hunk 1 above)
Comment (StephanCoetzee) [reply cpu_map_keyme2560.h:74]: The STEP and DIRECTION defines logically go together. In https://github.com/keyme/grbl/pull/112, I put the MOTOR_RESET defines between STEP and DIRECTION. This is a bit easier to follow IMO.

File: motor_driver.c (hunk 4 above)
Comment (StephanCoetzee) [reply motor_driver.c]: :+1:

**Comment by StephanCoetzee** [comment]
Updated and tested.

Also added a fix for the micro-step mask reading.

File: motor_driver.c [hunk 6]
```diff
@@ -142,4 +175,40 @@ void motor_drv_init()
...
+  motor_drv_set_isgain(XTABLE, FIVE);
+
+  motor_drv_set_torque(YTABLE, TORQUE_VAL_5A);
+  motor_drv_set_isgain(YTABLE, FIVE);
+
+  motor_drv_set_torque(CAROUSEL, TORQUE_VAL_5A);
+  motor_drv_set_isgain(CAROUSEL, FIVE);
+
+  motor_drv_set_torque(GRIPPER, TORQUE_VAL_3A);
+  motor_drv_set_isgain(GRIPPER, FIVE);
```
Comment (Jeff-Ciesielski) [inline motor_driver.c:198]: Food for thought (no need to change right now), but perhaps we should make this configurable in the future.

File: motor_driver.c [hunk 7]
```diff
@@ -20,64 +21,81 @@
...
+      I = 2.97
+*/
+
 
 const char * reg_names[] = {"CTRL", "TORQUE", "OFF", "BLANK",
                             "DECAY", "STALL", "DRIVE", "STATUS"};
 
-void _motor_drv_write_reg(enum stepper_e stepper, enum address_e address, uint8_t * data)
+void _motor_drv_write_reg(enum stepper_e stepper, enum address_e address, uint16_t data)
 {
```
Comment (Jeff-Ciesielski) [inline motor_driver.c:54]: One minor request: Can you leave a comment with the command format?

**Review by Jeff-Ciesielski — COMMENTED** [review]
One minor nit, otherwise LGTM

**Comment by StephanCoetzee** [comment]
Updated and tested.

File: motor_driver.c (hunk 7 above)
Comment (StephanCoetzee) [reply motor_driver.c:54]: addressed

**Comment by Jeff-Ciesielski** [comment]
LGTM
